pkg/monitor/lease.go

```diff
@@ -86,16 +122,204 @@ func leaseVIP(cfgPath, masterDevice, name string) error {
-	mac := GenerateMac(MacPrefixQumranet[:], name)
+func createLeaseFileWatcher(log logrus.FieldLogger, fileName string) (*fsnotify.Watcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.WithError(err).Error("Failed to add a create a new watcher")
+		return nil, err
+	}
 
+	err = watcher.Add(fileName)
+	if err != nil {
+		log.WithFields(logrus.Fields{
+			"filename": fileName,
+		}).WithError(err).Error("Failed to add a watcher to lease file")
+		return nil, err
+	}
+
+	return watcher, nil
+}
+
+func RunFiniteWatcher(log logrus.FieldLogger, watcher *fsnotify.Watcher, fileName, expectedIface, expectedIp string, write chan<- error) {
+	go func() {
+		defer watcher.Close()
+		done := false
+		var err error
+
+		for !done {
+			done, err = runWatcher(log, watcher, fileName, expectedIface, expectedIp)
+		}
+
+		write <- err
+	}()
+}
+
+func RunInfiniteWatcher(log logrus.FieldLogger, watcher *fsnotify.Watcher, fileName, expectedIface, expectedIp string) {
+	go func() {
+		defer watcher.Close()
+
+		for {
+			_, _ = runWatcher(log, watcher, fileName, expectedIface, expectedIp)
+		}
+	}()
+}
+
+func runWatcher(log logrus.FieldLogger, watcher *fsnotify.Watcher, fileName, expectedIface, expectedIp string) (bool, error) {
+	select {
+	case event, ok := <-watcher.Events:
+		if !ok {
+			return false, nil
+		}
+
+		if event.Op&fsnotify.Write == fsnotify.Write {
+			if stat, err := os.Stat(fileName); err != nil {
+				log.WithFields(logrus.Fields{
+					"filename": fileName,
+				}).WithError(err).Error("Failed to stat the leasing file")
+				return true, err
+			} else if stat.Size() == 0 {
+				// The file has been modified and truncated
+				return false, nil
+			}
+
+			if iface, ip, err := GetLastLeaseFromFile(log, fileName); err != nil {
+				log.WithFields(logrus.Fields{
+					"filename": fileName,
+				}).WithError(err).Error("Failed to get lease information from leasing file")
+				return true, err
+			} else if iface != expectedIface || (expectedIp != "" && ip != expectedIp) {
+				err := fmt.Errorf("A new lease has been written to the lease file with wrong data")
+				log.WithFields(logrus.Fields{
+					"filename":      fileName,
+					"iface":         iface,
+					"expectedIface": expectedIface,
+					"ip":            ip,
+					"expectedIp":    expectedIp,
+				}).Error(err)
+				return true, err
+			} else {
+				log.WithFields(logrus.Fields{
+					"fileName": fileName,
+					"iface":    iface,
+					"ip":       ip,
+				}).Info("A new lease has been written to the lease file with the right data")
+				return true, nil
+			}
+		}
+	case err, ok := <-watcher.Errors:
+		if !ok {
+			return false, nil
+		}
+
+		log.WithFields(logrus.Fields{
+			"filename": fileName,
+		}).WithError(err).Error("Lease file watcher error")
+		return true, err
+	}
+
+	return false, nil
+}
+
+func GetLastLeaseFromFile(log logrus.FieldLogger, fileName string) (string, string, error) {
+	data, err := ioutil.ReadFile(fileName)
```

Comment:
I think the only common functions should be GetLastLeaseFromFile, and LeaseInterface.  Otherwise, it will make both implementations more complicated that if implemented separately.

Reply:
I agree.